File: src/main/java/xyz/pupbrained/DropConfirm.java
package xyz.pupbrained;

import net.fabricmc.api.ClientModInitializer;
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents;
import net.fabricmc.fabric.api.client.keybinding.v1.KeyBindingHelper;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.option.KeyBinding;
import net.minecraft.sound.SoundEvents;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.lwjgl.glfw.GLFW;
import xyz.pupbrained.config.DropConfirmConfig;

import java.util.Objects;

public class DropConfirm implements ClientModInitializer {
  public static final Logger LOGGER = LogManager.getLogger("DropConfirm");

  @Override
  public void onInitializeClient() {
    DropConfirmConfig.GSON.load();

    var toggleKey = KeyBindingHelper.registerKeyBinding(new KeyBinding(
        "key.drop_confirm.toggle",
        GLFW.GLFW_KEY_UNKNOWN,
        "category.drop_confirm.keybinds"));

    ClientTickEvents.END_CLIENT_TICK.register(client -> {
      while (toggleKey.wasPressed()) {
        var mc = MinecraftClient.getInstance();
        var config = DropConfirmConfig.GSON.instance();
        var player = Objects.requireNonNull(mc.player);

        config.enabled = !config.enabled;

        DropConfirmConfig.GSON.save();

        if (config.playSounds)
          player.playSound(SoundEvents.ENTITY_ITEM_PICKUP, 1.0f, config.enabled ? 1.0f : 0.5f);

        mc.inGameHud.setOverlayMessage(
            Text
                .literal("DropConfirm: ")
                .append(
                    Text
                        .translatable(config.enabled ? "drop_confirm.toggle.on" : "drop_confirm.toggle.off")
                        .formatted(config.enabled ? Formatting.GREEN : Formatting.RED)),
            false);
      }
    });
  }
}

File: common/src/main/kotlin/xyz/pupbrained/drop_confirm/DropConfirm.kt
package xyz.pupbrained.drop_confirm

import com.mojang.blaze3d.platform.InputConstants
import net.minecraft.ChatFormatting
import net.minecraft.client.KeyMapping
import net.minecraft.client.Minecraft
import net.minecraft.network.chat.Component
import net.minecraft.sounds.SoundEvents
import org.lwjgl.glfw.GLFW

object DropConfirm {
  var isConfirmed = false

  val TOGGLE_KEY = KeyMapping(
    "key.drop_confirm.toggle",
    InputConstants.Type.KEYSYM,
    GLFW.GLFW_KEY_UNKNOWN,
    "category.drop_confirm.keybinds"
  )

  fun handleKeyPresses(mc: Minecraft) {
    while (TOGGLE_KEY.consumeClick()) {
      val config = DropConfirmConfig.GSON.instance()
      val player = mc.player ?: return

      (!config.enabled).also { config.enabled = it }
      DropConfirmConfig.GSON.save()

      if (config.playSounds)
        player.playSound(
          SoundEvents.ITEM_PICKUP,
          1.0f,
          if (config.enabled) 1.0f else 0.5f
        )

      mc.gui.setOverlayMessage(
        Component.literal("DropConfirm: ").append(
          Component.translatable(
            if (config.enabled) "drop_confirm.toggle.on" else "drop_confirm.toggle.off"
          ).withStyle(
            if (config.enabled) ChatFormatting.GREEN else ChatFormatting.RED
          )
        ),
        false
      )
    }
  }

  fun init() {
    DropConfirmConfig.GSON.load()
  }
}
